measureme/src/counters.rs

```diff
@@ -0,0 +1,869 @@
+                        (6, 61) => (Well, "Broadwell (U/Y/S)"),
+                        (6, 63) => (Well, "Haswell (E/EP/EX)"),
+                        (6, 69) => (Well, "Haswell (ULT)"),
+                        (6, 70) => (Well, "Haswell (GT3e)"),
+                        (6, 71) => (Well, "Broadwell (H/C/W)"),
+                        (6, 79) => (Well, "Broadwell (E/EP/EX)"),
+                        (6, 86) => (Well, "Broadwell (DE/Hewitt Lake)"),
+
+                        (6, 78) => (Lake, "Skylake (Y/U)"),
+                        (6, 85) => (Lake, "Skylake (SP/X/DE/W) / Cascade Lake (SP/X/W)"),
+                        (6, 94) => (Lake, "Skylake (DT/H/S)"),
+                        (6, 142) => (Lake, "Kaby Lake (Y/U) / Coffee Lake (U)"),
+                        (6, 158) => (Lake, "Kaby Lake (DT/H/S/X) / Coffee Lake (S/H/E)"),
+
+                        (6..=14, _) | (16..=0xffff_ffff, _) => {
+                            really_warn!(
+                                "CpuModel::detect: unknown Intel CPU (Family {} Model {}), \
+                                 assuming Skylake-like; {}",
+                                family,
+                                model,
+                                super::BUG_REPORT_MSG
+                            );
+
+                            (UnknownMaybeLakeLike, "")
+                        }
+                    };
+
+                    if !name.is_empty() {
+                        info!("CpuModel::detect: known Intel CPU: {}", name);
+                    }
+
+                    Ok(CpuModel::Intel(gen))
+                }
+
+                _ => Err(format!(
+                    "cpuid returned unknown CPU vendor {:?}; version={:#x}",
+                    vendor, version
+                )
+                .into()),
+            }
+        }
+
+        /// Return the hardware performance counter configuration for
+        /// counting "hardware interrupts" (documented or not).
+        fn irqs_counter_config(&self) -> Result<u32, Box<dyn Error + Send + Sync>> {
+            match self {
+                CpuModel::Amd(model) => match model {
+                    AmdGen::PreZen => Ok(0x00_cf),
+                    AmdGen::Zen | AmdGen::UnknownMaybeZenLike => Ok(0x00_2c),
+                },
+                CpuModel::Intel(model) => match model {
+                    IntelGen::PreBridge => Err(format!(
+                        "counting IRQs not yet supported on Intel CPUs \
+                         predating Sandy Bridge; {}",
+                        super::BUG_REPORT_MSG
+                    )
+                    .into()),
+                    IntelGen::Bridge
+                    | IntelGen::Well
+                    | IntelGen::Lake
+                    | IntelGen::UnknownMaybeLakeLike => Ok(0x01_cb),
+                },
+            }
+        }
+    }
+}
+
+#[cfg(not(all(feature = "nightly", target_arch = "x86_64", target_os = "linux")))]
```

Comment:
Note to me: We should enable the nightly feature flag on our nightly toolchain CI build.